src/adapter/src/coord/cluster_scheduling.rs

```diff
@@ -0,0 +1,169 @@
+                                    .write_frontier;
+                                mv_frontier.less_than(&local_read_ts)
+                            }),
+                        ));
+                    }
+                }
+            }
+        }
+        decisions
+    }
+
+    #[mz_ore::instrument(level = "debug")]
+    pub(crate) async fn handle_scheduling_decisions(
+        &mut self,
+        decisions: Vec<(&'static str, Vec<(ClusterId, bool)>)>,
+    ) {
+        for (policy_name, decisions) in decisions.iter() {
+            for (cluster_id, decision) in decisions {
+                self.scheduling_decisions
+                    .entry(*cluster_id)
+                    .or_insert_with(Default::default)
+                    .insert(policy_name, *decision);
+            }
+        }
+
+        // Clean up those clusters from `scheduling_decisions` that
+        // - have been dropped, or
+        // - were switched to unmanaged, or
+        // - were switched to `SCHEDULE = MANUAL`.
+        for cluster_id in self.scheduling_decisions.keys().cloned().collect_vec() {
+            match self.get_managed_cluster_config(cluster_id) {
+                None => {
+                    // Cluster have been dropped or switched to unmanaged.
+                    self.scheduling_decisions.remove(&cluster_id);
+                }
+                Some(managed_config) => {
+                    if matches!(managed_config.schedule, ClusterScheduleOptionValue::Manual) {
+                        self.scheduling_decisions.remove(&cluster_id);
+                    }
+                }
+            }
+        }
+
+        for (cluster_id, decisions) in self.scheduling_decisions.clone() {
+            // If all policies have made a decision about this cluster
+            if POLICIES.iter().all(|policy| decisions.contains_key(policy)) {
+                // check whether the cluster's state matches the needed state.
+                let needs_replica = decisions.values().contains(&true);
+                let cluster_config = self
+                    .get_managed_cluster_config(cluster_id)
+                    .expect("cleaned up non-existing and unmanaged clusters above");
+                let has_replica = cluster_config.replication_factor > 0; // Is it on?
+                if needs_replica != has_replica {
+                    // Turn the cluster on or off.
+                    let mut new_config = cluster_config.clone();
+                    new_config.replication_factor = if needs_replica { 1 } else { 0 };
+                    if let Err(e) = self
+                        .sequence_alter_cluster_managed_to_managed(
+                            None,
+                            cluster_id,
+                            &cluster_config,
+                            new_config.clone(),
+                        )
+                        .await
```

Comment:
What user does this appear as in `mz_audit_events`? `mz_system`, I hope? If not, we should fix it!

Reply:
The `user` column is `null` for these events in `mz_audit_events`. Is that ok? E.g. the following is after locally starting envd, creating a few objects, and then setting a cluster to `SCHEDULE = REFRESH`, and waiting a bit:
```
 id | event_type |    object_type    |                                                                         details                                                                         |    user     |        occurred_at         
----+------------+-------------------+---------------------------------------------------------------------------------------------------------------------------------------------------------+-------------+----------------------------
 30 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","replica_id":"u7","replica_name":"r1"}         |             | 2024-04-05 12:38:07.765+00
 29 | drop       | cluster-replica   | {"cluster_id":"u2","cluster_name":"c1","replica_id":"u6","replica_name":"r1"}                                                                           |             | 2024-04-05 12:37:49.706+00
 28 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","replica_id":"u6","replica_name":"r1"}         |             | 2024-04-05 12:37:46.763+00
 27 | drop       | cluster-replica   | {"cluster_id":"u2","cluster_name":"c1","replica_id":"u5","replica_name":"r1"}                                                                           |             | 2024-04-05 12:37:31.709+00
 26 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","replica_id":"u5","replica_name":"r1"}         |             | 2024-04-05 12:37:28.754+00
 25 | drop       | cluster-replica   | {"cluster_id":"u2","cluster_name":"c1","replica_id":"u4","replica_name":"r1"}                                                                           |             | 2024-04-05 12:37:10.709+00
 24 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","replica_id":"u4","replica_name":"r1"}         |             | 2024-04-05 12:37:07.764+00
 23 | drop       | cluster-replica   | {"cluster_id":"u2","cluster_name":"c1","replica_id":"u3","replica_name":"r1"}                                                                           |             | 2024-04-05 12:36:49.706+00
 22 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","replica_id":"u3","replica_name":"r1"}         |             | 2024-04-05 12:36:46.773+00
 21 | drop       | cluster-replica   | {"cluster_id":"u2","cluster_name":"c1","replica_id":"u2","replica_name":"r1"}                                                                           |             | 2024-04-05 12:36:31.705+00
 20 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u2","cluster_name":"c1","disk":false,"internal":false,"logical_size":"1","replica_id":"u2","replica_name":"r1"}         |             | 2024-04-05 12:36:28.757+00
 19 | create     | materialized-view | {"database":"materialize","id":"u2","item":"mv1","schema":"public"}                                                                                     | materialize | 2024-04-05 12:36:25.763+00
 18 | create     | cluster           | {"id":"u2","name":"c1"}                                                                                                                                 | materialize | 2024-04-05 12:36:25.483+00
 17 | create     | table             | {"database":"materialize","id":"u1","item":"t","schema":"public"}                                                                                       | materialize | 2024-04-05 12:36:25.15+00
 1  | create     | role              | {"id":"u1","name":"materialize"}                                                                                                                        |             | 2024-04-05 12:36:10.082+00
 11 | create     | cluster           | {"id":"u1","name":"quickstart"}                                                                                                                         |             | 2024-04-05 12:36:10.082+00
 6  | create     | database          | {"id":"u1","name":"materialize"}                                                                                                                        |             | 2024-04-05 12:36:10.082+00
 9  | create     | schema            | {"database_name":"materialize","id":"3","name":"public"}                                                                                                |             | 2024-04-05 12:36:10.082+00
 12 | grant      | cluster           | {"grantee_id":"p","grantor_id":"s1","object_id":"Cu1","privileges":"U"}                                                                                 |             | 2024-04-05 12:36:10.082+00
 7  | grant      | database          | {"grantee_id":"p","grantor_id":"s1","object_id":"Du1","privileges":"U"}                                                                                 |             | 2024-04-05 12:36:10.082+00
 13 | grant      | cluster           | {"grantee_id":"u1","grantor_id":"s1","object_id":"Cu1","privileges":"UC"}                                                                               |             | 2024-04-05 12:36:10.082+00
 5  | grant      | type              | {"database_id":null,"grantee_id":"p","privileges":"U","role_id":"p","schema_id":null}                                                                   |             | 2024-04-05 12:36:10.082+00
 8  | grant      | database          | {"grantee_id":"u1","grantor_id":"s1","object_id":"Du1","privileges":"UC"}                                                                               |             | 2024-04-05 12:36:10.082+00
 10 | grant      | schema            | {"grantee_id":"u1","grantor_id":"s1","object_id":"Su1.u3","privileges":"UC"}                                                                            |             | 2024-04-05 12:36:10.082+00
 15 | grant      | system            | {"grantee_id":"s1","grantor_id":"s1","object_id":"SYSTEM","privileges":"RBN"}                                                                           |             | 2024-04-05 12:36:10.082+00
 16 | grant      | system            | {"grantee_id":"u1","grantor_id":"s1","object_id":"SYSTEM","privileges":"RBN"}                                                                           |             | 2024-04-05 12:36:10.082+00
 4  | grant      | schema            | {"database_id":null,"grantee_id":"s2","privileges":"U","role_id":"p","schema_id":null}                                                                  |             | 2024-04-05 12:36:10.082+00
 2  | grant      | cluster           | {"database_id":null,"grantee_id":"s2","privileges":"U","role_id":"p","schema_id":null}                                                                  |             | 2024-04-05 12:36:10.082+00
 3  | grant      | database          | {"database_id":null,"grantee_id":"s2","privileges":"U","role_id":"p","schema_id":null}                                                                  |             | 2024-04-05 12:36:10.082+00
 14 | create     | cluster-replica   | {"billed_as":null,"cluster_id":"u1","cluster_name":"quickstart","disk":false,"internal":false,"logical_size":"1","replica_id":"u1","replica_name":"r1"} |             | 2024-04-05 12:36:10.082+00
(30 rows)
```

Reply:
Changing this to `mz_system` might go in a follow-up PR, see Parker's [comment](https://github.com/MaterializeInc/materialize/pull/26401#discussion_r1553731534).